### PR TITLE
Bubble and insertion sort in Dart

### LINKED ISSUE #490: Adding Algorithms in Dart & Kotlin & Rust
I would like to propose 2 new languages in which Algorithms can added.
- Dart
- Kotlin
- Rust
 
I'm GSSoC 19 Participant, since there was no issue for it, I created this one. Let me know your thoughts.

### PR DESCRIPTION
**Pull request by AadumKhor**
Fixes #490 

#### Checklist:
- [X] 4 space indentation.
- [X] Coding conventions are followed.
- [x] Sample Input / Output.
- [X] Logic Documentation (Comments).
- [X] File names are correct.

#### Changes proposed in this pull request:

-  Dart sorting algorithms added 

#### Languages Used:
- Dart
#### Files Added:
- Bubble_Sort.dart
- Insertion_Sort.dart
- Merge_Sort.dart
- Heap_Sort.dart
- Quick_Sort.dart
- Activity_Selection.dart

### CODE REVIEW

**Comment by AadumKhor** [comment]
Once this is accepted I can gradually add other algorithms in Dart as well.

File: Bubble_Sort/bubble_sort.dart [hunk 1]
```diff
@@ -0,0 +1,56 @@
+import 'dart:io';
+
+class BubbleSort {
+  List<int> unsorted;
+  BubbleSort(this.unsorted);
+
+  // *add array that you want to sort
```
Comment (MastersAbh) [inline Bubble_Sort/bubble_sort.dart]: remove * symbol from comment.

File: Bubble_Sort/bubble_sort.dart [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+  var countString = stdin.readLineSync();
+  int count = int.parse(countString);
+  bubbleSort.addArray(count, unsorted);
+  print('The given array is : ');
+  bubbleSort.printArray(bubbleSort.unsorted);
+  // sort the array and display using the next print
+  bubbleSort.sort(bubbleSort.unsorted);
+  print('Sorted Array is:');
+  bubbleSort.printArray(bubbleSort.unsorted);
+}
```
Comment (MastersAbh) [inline Bubble_Sort/bubble_sort.dart]: add a blank line at the end of the file.

File: Insertion_Sort/insertion_sort.dart [hunk 3]
```diff
@@ -0,0 +1,52 @@
+import 'dart:io';
+
+class InsertionSort {
+  List<int> array;
+
+  InsertionSort(this.array);
+
+  void insertionSort(List<int> a, int size) {
+    int temp, j;
+
```
Comment (MastersAbh) [inline Insertion_Sort/insertion_sort.dart]: Remove extra blank line

File: Insertion_Sort/insertion_sort.dart [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    for (int i = 1; i < a.length; i++) {
+      temp = a[i];
+      j = i - 1;
+
```
Comment (MastersAbh) [inline Insertion_Sort/insertion_sort.dart]: Remove extra blank line

File: Insertion_Sort/insertion_sort.dart [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+      while (j >= 0 && array[j] > temp) {
+        a[j + 1] = a[j];
+        j--;
+      }
+
```
Comment (MastersAbh) [inline Insertion_Sort/insertion_sort.dart]: Remove extra blank line

**Comment by jainaman224** [comment]
Please fix file name.

**Comment by somya-kapoor** [comment]
Please add sample input and output.

File: Bubble_Sort/bubble_sort.dart [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+      bool swaped = false;
+      for (var j = i + 1; j < a.length; j++) {
+        if (a[i] > a[j]) {
+          var tmp = a[i];
+          a[i] = a[j];
+          a[j] = tmp;
+          swaped = true;
+        }
+      }
+      if (!swaped) break;
```
Comment (somya-kapoor) [inline Bubble_Sort/bubble_sort.dart]: Please fix the indentation

**Comment by AadumKhor** [comment]
I'll submit it with all the corrections by afternoon. Thank you for the feedback.

**Comment by AadumKhor** [comment]
> 
> 
> Please fix file name.

Which file exactly?

**Comment by MastersAbh** [comment]
> > Please fix file name.
> 
> Which file exactly?

Use Bubble_Sort.dart instead of bubble_sort.dart

Similarly for Insertion Sort.

**Comment by AadumKhor** [comment]
I have committed the changes. Please check and update.

**Comment by AadumKhor** [comment]
image
I changed it to 4 now. I hope it's okay.

File: Bubble_Sort/bubble_sort.dart [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+
+class BubbleSort 
+{
+    List<int> unsorted;
+    BubbleSort(this.unsorted);
+
+    //add array that you want to sort
+    void addArray(int count, List<int> list) 
+    {
+      list = [];
```
Comment (MastersAbh) [inline Bubble_Sort/bubble_sort.dart]: this line should also have 4 spaces under loop

File: Bubble_Sort/bubble_sort.dart [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+      unsorted = list;
```
Comment (MastersAbh) [inline Bubble_Sort/bubble_sort.dart]: this too

File: Heap_Sort/Heap_Sort.dart [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+    {
+        for (var i = 0; i < a.length; i++) 
+        {
+            print(a[i]);
+        }
+    }
+
+    void heap(List list, int n, int i) 
+    {
+      int largest = i;
```
Comment (MastersAbh) [inline Heap_Sort/Heap_Sort.dart]: 4 spaces

File: Heap_Sort/Heap_Sort.dart [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+      int l = 2 * i + 1;
```
Comment (MastersAbh) [inline Heap_Sort/Heap_Sort.dart]: 4 spaces

File: Heap_Sort/Heap_Sort.dart [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+      int r = 2 * i + 2;
```
Comment (MastersAbh) [inline Heap_Sort/Heap_Sort.dart]: 4 spaces

File: Heap_Sort/Heap_Sort.dart [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+            swapList(list, i, largest);
+            heap(list, n, largest);
+        }
+    }
+
+    void swapList(List list, int i, int largest) 
+    {
+      int swap = list[i];
+      list[i] = list[largest];
+      list[largest] = swap;
```
Comment (MastersAbh) [inline Heap_Sort/Heap_Sort.dart]: 4 spaces indentation inside the function

File: Heap_Sort/Heap_Sort.dart [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+    }
+
+    void heapSort(List list) 
+    {
+      int n = list.length;
```
Comment (MastersAbh) [inline Heap_Sort/Heap_Sort.dart]: 4 spaces

File: Heap_Sort/Heap_Sort.dart [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+            swap(list, i);
+            heap(list, i, 0);
+        }
+    }
+
+    void swap(List list, int i) 
+    {
+      int temp = list[0];
+      list[0] = list[i];
+      list[i] = temp;
```
Comment (MastersAbh) [inline Heap_Sort/Heap_Sort.dart]: 4 spaces indentation

File: Insertion_Sort/insertion_sort.dart [hunk 15]
```diff
@@ -0,0 +1,79 @@
+import 'dart:io';
+class InsertionSort 
+{
+    List<int> array;
+    InsertionSort(this.array);
+
+    void insertionSort(List<int> a, int size) 
+    {
+      int temp, j;
```
Comment (MastersAbh) [inline Insertion_Sort/insertion_sort.dart]: 4 space

File: Merge_Sort/Merge_Sort.dart [hunk 16]
```diff
@@ -0,0 +1,119 @@
...
+      List leftList = new List(leftSize); // make the array
+      List rightList = new List(rightSize);
+
+      for (int i = 0; i < leftSize; i++) leftList[i] = list[leftIndex + i]; // making the arrays
+      for (int j = 0; j < rightSize; j++)
+        rightList[j] = list[middleIndex + j + 1];
+
+      int i = 0, j = 0;
+      int k = leftIndex;
+
```
Comment (MastersAbh) [inline Merge_Sort/Merge_Sort.dart]: 4 spaces indentation in the function.

File: Merge_Sort/Merge_Sort.dart [hunk 17]
```diff
@@ -0,0 +1,119 @@
...
+      while (i < leftSize && j < rightSize) 
+      {
+            if (leftList[i] <= rightList[j])
```
Comment (MastersAbh) [inline Merge_Sort/Merge_Sort.dart]: here, there are 5 spaces instead of 4

**Comment by AadumKhor** [comment]
Did it manually now. Please check or else recommend settings for VSCode to help me with this.

**Comment by AadumKhor** [comment]
The tab is set to give 4 spaces in my settings. I checked that. I will make the changes in the required file.

File: Activity_Selection/Activity_Selection.dart [hunk 18]
```diff
@@ -0,0 +1,38 @@
+/* The Activity Selection Problem
+You are given n activities with their start and finish times. 
+Select the maximum number of activities that can be performed by a single person, 
+assuming that a person can only work on a single activity at a time. 
+*/
+class ActivitySelection 
+{
+    void activities(List<int> start, List<int> finish, int n)
```
Comment (MastersAbh) [inline Activity_Selection/Activity_Selection.dart]: The indentation of this function is more than 4 spaces. You can make this out by seeing it. 
Please fix this function. Rest is fine.

File: Activity_Selection/Activity_Selection.dart [hunk 19]
```diff
@@ -0,0 +1,38 @@
...
+    selection.activities(start, finish, n);
+}
+/*
+	Sample input
+	Start : 0 5 5 4 5 8
+	Finish: 2 7 6 8 9 14
+
+	Sample Output
+	0 1 5
+ */
```
Comment (MastersAbh) [inline Activity_Selection/Activity_Selection.dart]: A blank line at the end of the file.

File: Activity_Selection/Activity_Selection.dart (hunk 18 above)
Comment (AadumKhor) [reply Activity_Selection/Activity_Selection.dart]: image
The editor says it's fine.  I have added the blank line at the end.

Comment (MastersAbh) [reply Activity_Selection/Activity_Selection.dart]: Hi, I know it shows correct in your editor but it has different indentation from the rest of file here. 
Can you do one thing? 
Just manually delete the spaces infront of each line and add required no. of spaces : 4 per block and then commit.
I think then it should work. Most probably, tabs and spaces would have got mixed.

Comment (AadumKhor) [reply Activity_Selection/Activity_Selection.dart]: > 
> 
> Hi, I know it shows correct in your editor but it has different indentation from the rest of file here.
> Can you do one thing?
> Just manually delete the spaces infront of each line and add required no. of spaces : 4 per block and then commit.
> I think then it should work. Most probably, tabs and spaces would have got mixed.
Okay, on it.

Comment (AadumKhor) [reply Activity_Selection/Activity_Selection.dart]: I've made another commit. Please check

Comment (AadumKhor) [reply Activity_Selection/Activity_Selection.dart]: In the editor it is just as you want it to be. Even if I manually do 4 spaces it is the same thing as the commit I've already made. And the screenshot above shows it. On the github viewer however what you're saying is correct. The brackets are not in line. When I do the brackets again it reverts back to the original commit so basically no changes are made. Is there any other way? I have tried both tabs and spaces and both of them give the same result.

File: Activity_Selection/Activity_Selection.dart [hunk 20]
```diff
@@ -0,0 +1,38 @@
+/* The Activity Selection Problem
+You are given n activities with their start and finish times. 
+Select the maximum number of activities that can be performed by a single person, 
+assuming that a person can only work on a single activity at a time. 
+*/
+class ActivitySelection 
+{
+    void activities(List<int> start, List<int> finish, int n) 
+	{
```
[the remark this answers is not in the document]
Comment (AadumKhor) [reply Activity_Selection/Activity_Selection.dart]: Did the brackets manually now. Checked the file on github and it looks exactly as you described.

File: Activity_Selection/Activity_Selection.dart [hunk 21]
```diff
@@ -0,0 +1,38 @@
+/* The Activity Selection Problem
+You are given n activities with their start and finish times. 
+Select the maximum number of activities that can be performed by a single person, 
+assuming that a person can only work on a single activity at a time. 
+*/
+class ActivitySelection 
+{
+    void activities(List<int> start, List<int> finish, int n) 
+    {
+	    int i = 0;
```
Comment (MastersAbh) [inline Activity_Selection/Activity_Selection.dart]: Line 10 is TAB + 4 spaces.
Make it 8 spaces.

File: Activity_Selection/Activity_Selection.dart [hunk 22]
```diff
@@ -0,0 +1,38 @@
...
+	    print('$i ');
+	    for (int j = 1; j < n; j++) 
+	    {
```
Comment (MastersAbh) [inline Activity_Selection/Activity_Selection.dart]: Line 11 to 13 are TAB + 4 spaces.
Make them 8 spaces.

File: Activity_Selection/Activity_Selection.dart [hunk 23]
```diff
@@ -0,0 +1,38 @@
...
+		    if (start[j] >= finish[i]) 
+		    {
```
Comment (MastersAbh) [inline Activity_Selection/Activity_Selection.dart]: line 14 to 15 are TAB+TAB+4 spaces.
Make it 12 spaces

File: Activity_Selection/Activity_Selection.dart [hunk 24]
```diff
@@ -0,0 +1,38 @@
...
+			    print('$j ');
+			    i = j;
```
Comment (MastersAbh) [inline Activity_Selection/Activity_Selection.dart]: 16 - 17 are TAB+TAB+TAB+4 spaces.
Make it 16 spaces.

File: Activity_Selection/Activity_Selection.dart [hunk 25]
```diff
@@ -0,0 +1,38 @@
...
+		    }
+	    }
```
Comment (MastersAbh) [inline Activity_Selection/Activity_Selection.dart]: 19 Should be 8 spaces

File: Activity_Selection/Activity_Selection.dart [hunk 26]
```diff
@@ -0,0 +1,38 @@
...
+    {
+	    int i = 0;
+	    print('$i ');
+	    for (int j = 1; j < n; j++) 
+	    {
+		    if (start[j] >= finish[i]) 
+		    {
+			    print('$j ');
+			    i = j;
+		    }
```
Comment (MastersAbh) [inline Activity_Selection/Activity_Selection.dart]: 18 should be 12 spaces.

File: Bubble_Sort/bubble_sort.dart [hunk 27]
```diff
@@ -0,0 +1,96 @@
...
+class BubbleSort 
+{
+    List<int> unsorted;
+    BubbleSort(this.unsorted);
+
+    //add array that you want to sort
+    void addArray(int count, List<int> list) 
+    {
+		list = [];
+		unsorted = list;
```
Comment (MastersAbh) [inline Bubble_Sort/bubble_sort.dart]: Something happened to line 11, 12.
Tabs ? Please fix this.

Comment (AadumKhor) [reply Bubble_Sort/bubble_sort.dart]: Done. Please check

File: Quick_Sort/Quick_Sort.dart [hunk 28]
```diff
@@ -0,0 +1,101 @@
...
+        int pivot = list[low];
+        int i = high;
+        int temp;
+        for (int j = high; j > low; j--) 
+        {
+            if (list[j] > pivot) {
+            temp = list[j];
+            list[j] = list[i];
+            list[i] = temp;
+            i--;
```
Comment (MastersAbh) [inline Quick_Sort/Quick_Sort.dart]: Indentation of 35 to 38

File: Activity_Selection/Activity_Selection.dart [hunk 29]
```diff
@@ -0,0 +1,38 @@
...
+                i = j;
+            }
+        }
+    }
+}
+
+void main() 
+{
+    List<int> start = [0, 5, 5, 4, 5, 8]; // starting time of the activities
+    List<int> finish = [2,7,6,8,9,14]; // finishin time of the activities
```
Comment (MastersAbh) [inline Activity_Selection/Activity_Selection.dart]: Add a space after comma inside the list.

File: Quick_Sort/Quick_Sort.dart (hunk 28 above)
Comment (AadumKhor) [reply Quick_Sort/Quick_Sort.dart]: There is nothing to fix. It is 12 spaces.

Comment (MastersAbh) [reply Quick_Sort/Quick_Sort.dart]: Should be 16. It is inside if condition.

**Comment by AadumKhor** [comment]
> 
> 
> After doing these changes, do squash your commits.

How do you squash commits?

File: Quick_Sort/Quick_Sort.dart (hunk 28 above)
Comment (MastersAbh) [reply Quick_Sort/Quick_Sort.dart]: Please fix this

Comment (AadumKhor) [reply Quick_Sort/Quick_Sort.dart]: Done.

**Comment by MastersAbh** [comment]
> > After doing these changes, do squash your commits.
> 
> How do you squash commits?

Can we connect on slack over this ?

**Comment by MastersAbh** [comment]
Do
git reset --soft HEAD~N
where N is the number of commits. In your case, 13.
then git commit
then git push origin -u -f Branchname

**Comment by AadumKhor** [comment]
Thank you for the help.

**Comment by jainaman224** [comment]
@MastersAbh Please review.

**Comment by MastersAbh** [comment]
You're welcome @AadumKhor .
It is good to go now. I approve this @jainaman224

**Pull request merged**